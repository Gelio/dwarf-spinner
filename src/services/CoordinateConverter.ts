import { Vec3 } from 'cannon';
import { vec3 } from 'gl-matrix';

// tslint:disable-next-line:no-stateless-class
export class CoordinateConverter {
  public static renderingToPhysics(v: vec3): Vec3 {
    return new Vec3(-v[0], v[2], v[1]);
  }

  public static physicsToRendering(out: vec3, source: Vec3): void;
  public static physicsToRendering(v: Vec3): vec3;
  public static physicsToRendering(
    vOrOut: Vec3 | vec3,
    source?: Vec3
  ): vec3 | void {
    if (source) {
      const out = <vec3>vOrOut;
      out[0] = -source.x;
      out[1] = source.z;
      out[2] = source.y;
    } else {
      const v = <Vec3>vOrOut;

      return vec3.fromValues(-v.x, v.z, v.y);
    }
  }
}
