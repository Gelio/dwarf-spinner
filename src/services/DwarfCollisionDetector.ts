import { ApplicationWorld } from 'models/ApplicationWorld';

import { GameStateType } from 'common/GameStateType';

import { ApplicationEventEmitter } from 'events/ApplicationEventEmitter';
import { DwarfGroundCollisionEvent } from 'events/DwarfGroundCollisionEvent';
import { RestartEvent } from 'events/RestartEvent';

import { changeGameState } from 'actions/GameActions';
import { store } from 'store';

export class DwarfCollisionDetector {
  private readonly eventEmitter: ApplicationEventEmitter;
  private readonly applicationWorld: ApplicationWorld;

  private dwarfCollided = false;

  constructor(eventEmitter: ApplicationEventEmitter, applicationWorld: ApplicationWorld) {
    this.eventEmitter = eventEmitter;
    this.applicationWorld = applicationWorld;

    this.onGroundCollision = this.onGroundCollision.bind(this);
    this.onRestartEvent = this.onRestartEvent.bind(this);
  }

  public init() {
    this.applicationWorld.groundBody.addEventListener('collide', this.onGroundCollision);
    this.eventEmitter.on(RestartEvent.name, this.onRestartEvent);
  }

  public destroy() {
    this.applicationWorld.groundBody.removeEventListener('collide', this.onGroundCollision);
    this.eventEmitter.removeListener(RestartEvent.name, this.onRestartEvent);
  }

  private onGroundCollision(event: any) {
    if (event.body !== this.applicationWorld.dwarf.body) {
      return;
    }

    if (!this.dwarfCollided) {
      store.dispatch(changeGameState(GameStateType.DwarfLanded));
    }

    this.eventEmitter.emitAppEvent(new DwarfGroundCollisionEvent(!this.dwarfCollided));
    this.dwarfCollided = true;
  }

  private onRestartEvent() {
    this.dwarfCollided = false;
  }
}
