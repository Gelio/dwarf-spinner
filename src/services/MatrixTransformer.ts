import { mat4 } from 'gl-matrix';

export class MatrixTransformer {
  public scale(matrix: mat4, vector: number[]) {
    mat4.scale(matrix, matrix, vector);
  }

  public translate(matrix: mat4, vector: number[]) {
    mat4.translate(matrix, matrix, vector);
  }

  public rotate(matrix: mat4, radians: number, axis: number[]) {
    mat4.rotate(matrix, matrix, radians, axis);
  }

  public identity(matrix: mat4) {
    mat4.identity(matrix);
  }
}
