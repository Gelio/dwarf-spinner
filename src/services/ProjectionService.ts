import { mat4 } from 'gl-matrix';

export class ProjectionService {
  public createProjectionMatrix(
    fovy: number,
    aspectRatio: number,
    near: number,
    far: number
  ): mat4 {
    const projectionMatrix = mat4.create();
    mat4.perspective(projectionMatrix, fovy, aspectRatio, near, far);

    return projectionMatrix;
  }
}
