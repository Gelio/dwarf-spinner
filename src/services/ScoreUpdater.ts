import { GameStateType } from 'common/GameStateType';
import { configuration } from 'configuration';

import { ApplicationEventEmitter } from 'events/ApplicationEventEmitter';
import { RestartEvent } from 'events/RestartEvent';

import { ApplicationWorld } from 'models/ApplicationWorld';

import { updateScore } from 'actions/GameActions';
import { getGameState, store } from 'store';

export class ScoreUpdater {
  private readonly applicationWorld: ApplicationWorld;
  private readonly eventEmitter: ApplicationEventEmitter;
  private intervalId: number;

  constructor(applicationWorld: ApplicationWorld, eventEmitter: ApplicationEventEmitter) {
    this.applicationWorld = applicationWorld;
    this.eventEmitter = eventEmitter;

    this.updateScore = this.updateScore.bind(this);
  }

  public init() {
    // @ts-ignore
    this.intervalId = setInterval(this.updateScore, configuration.scoreUpdateInterval);
    this.eventEmitter.on(RestartEvent.name, this.updateScore);
  }

  public destroy() {
    clearInterval(this.intervalId);
    this.eventEmitter.removeListener(RestartEvent.name, this.updateScore);
  }

  private updateScore() {
    if (getGameState() === GameStateType.AcceleratingSpinner) {
      return;
    }

    const newScore = this.getScore();
    const oldScore = store.getState().game.get('currentScore');

    if (Math.abs(newScore - oldScore) > configuration.scoreDifferenceThreshold) {
      store.dispatch(updateScore(this.getScore()));

    }
  }

  private getScore() {
    const dwarfPosition = this.applicationWorld.dwarf.body.position;
    const spinnerPosition = this.applicationWorld.fidgetSpinner.body.position;

    const xDelta = dwarfPosition.x - spinnerPosition.x;
    const yDelta = dwarfPosition.y - spinnerPosition.y;

    return Math.round(Math.sqrt(xDelta ** 2 + yDelta ** 2) * 10) / 10;
  }
}
