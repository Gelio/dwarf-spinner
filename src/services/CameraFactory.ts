import { Vec3 } from 'cannon';

import { ApplicationWorld } from 'models/ApplicationWorld';
import { GeneralCamera } from 'models/cameras/GeneralCamera';
import { ObservingCamera } from 'models/cameras/ObservingCamera';

import { Camera } from 'interfaces/Camera';

import { CoordinateConverter } from 'services/CoordinateConverter';

export class CameraFactory {
  private readonly applicationWorld: ApplicationWorld;
  private readonly stationaryPosition = new Vec3(-8, 0, 2);

  constructor(applicationWorld: ApplicationWorld) {
    this.applicationWorld = applicationWorld;
  }

  public createStationaryCamera(): Camera {
    const position = this.stationaryPosition.clone();
    const target = new Vec3(2, 2, 2);

    return new GeneralCamera(
      CoordinateConverter.physicsToRendering(position),
      CoordinateConverter.physicsToRendering(target)
    );
  }

  public createObservingCamera(): Camera {
    const position = this.stationaryPosition.clone();
    const dwarfBody = this.applicationWorld.dwarf.body;

    return new ObservingCamera(
      CoordinateConverter.physicsToRendering(position),
      dwarfBody
    );
  }
}
