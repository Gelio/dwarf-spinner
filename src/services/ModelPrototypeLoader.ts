// tslint:disable-next-line:no-require-imports
import expandVertexData = require('expand-vertex-data');

import { BufferUsageType } from 'common/BufferUsageType';
import { ModelPrototype } from 'models/ModelPrototype';
import { WebGLArrayBufferFacade } from 'models/WebGLArrayBufferFacade';
import { WebGLElementArrayBufferFacade } from 'models/WebGLElementArrayBufferFacade';
import { WebGLTextureFacade } from 'models/WebGLTextureFacade';
import { ImageLoader } from 'services/ImageLoader';

export class ModelPrototypeLoader {
  private nextTextureId = 0;
  private readonly gl: WebGLRenderingContext;
  private readonly imageLoader: ImageLoader;

  public constructor(gl: WebGLRenderingContext, imageLoader: ImageLoader) {
    this.gl = gl;
    this.imageLoader = imageLoader;
  }

  public async loadModelPrototype(modelObjUrl: string, textureImageUrl: string) {
    const [modelObjResponse, textureImage] = await Promise.all([
      fetch(modelObjUrl),
      this.imageLoader.loadImage(textureImageUrl)
    ]);
    const vertexData = await modelObjResponse.json();

    const gl = this.gl;

    const expandedVertexData = expandVertexData(vertexData, {
      facesToTriangles: true
    });

    const positionBuffer = this.createFloat32Buffer(
      gl,
      expandedVertexData.positions
    );
    const vertexNormalBuffer = this.createFloat32Buffer(
      gl,
      expandedVertexData.normals
    );
    const vertexTextureCoordsBuffer = this.createFloat32Buffer(
      gl,
      expandedVertexData.uvs,
      2
    );
    const positionIndicesBuffer = new WebGLElementArrayBufferFacade(
      gl,
      new Uint16Array(expandedVertexData.positionIndices),
      1,
      expandedVertexData.positionIndices.length,
      BufferUsageType.StaticDraw
    );

    const texture = new WebGLTextureFacade(
      gl,
      textureImage,
      this.nextTextureId
    );
    this.nextTextureId += 1;

    return new ModelPrototype(
      vertexNormalBuffer,
      vertexTextureCoordsBuffer,
      positionBuffer,
      positionIndicesBuffer,
      texture
    );
  }

  private createFloat32Buffer(gl: WebGLRenderingContext, data: number[], itemSize: number = 3) {
    return new WebGLArrayBufferFacade(
      gl,
      new Float32Array(data),
      itemSize,
      data.length / itemSize,
      BufferUsageType.StaticDraw
    );
  }
}
