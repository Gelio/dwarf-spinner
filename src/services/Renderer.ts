import { mat4 } from 'gl-matrix';

import { Camera } from 'interfaces/Camera';
import { Model } from 'interfaces/Model';

import { WebGLBinder } from 'services/WebGLBinder';

export class Renderer {
  private readonly canvas: HTMLCanvasElement;
  private readonly gl: WebGLRenderingContext;

  private readonly projectionMatrix: mat4;
  private readonly webGLBinder: WebGLBinder;

  private camera: Camera;

  public constructor(
    canvas: HTMLCanvasElement,
    gl: WebGLRenderingContext,
    projectionMatrix: mat4,
    webGLBinder: WebGLBinder
  ) {
    this.canvas = canvas;
    this.gl = gl;

    this.projectionMatrix = projectionMatrix;
    this.webGLBinder = webGLBinder;
  }

  public init() {
    const { gl, canvas } = this;

    gl.viewport(0, 0, canvas.width, canvas.height);
    gl.clearColor(0.0, 0.0, 0.0, 1.0);
    gl.enable(gl.DEPTH_TEST);
    this.clearCanvas();

    this.webGLBinder.bindProjectionMatrix(this.projectionMatrix);
  }

  public refreshCamera() {
    this.camera.update();
    this.webGLBinder.bindViewMatrix(this.camera.viewMatrix);
    this.webGLBinder.bindViewerPosition(this.camera.position);
  }

  public setActiveCamera(camera: Camera) {
    this.camera = camera;
  }

  public clearCanvas() {
    const gl = this.gl;

    // tslint:disable-next-line:no-bitwise
    gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);
  }

  public drawModel(model: Model) {
    model.draw(this.gl, this.webGLBinder);
  }
}
