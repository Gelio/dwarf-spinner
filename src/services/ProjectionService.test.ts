import { ProjectionService } from 'services/ProjectionService';

describe('ProjectionService', () => {
  describe('createProjectionMatrix', () => {
    it('should return a projection matrix', () => {
      const projectionService = new ProjectionService();

      const projectionMatrix = projectionService.createProjectionMatrix(45, 1.5, 0.1, 1000);
      expect(projectionMatrix).toMatchSnapshot();
    });
  });
});
