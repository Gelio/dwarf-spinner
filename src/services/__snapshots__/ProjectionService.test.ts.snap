// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectionService createProjectionMatrix should return a projection matrix 1`] = `
Float32Array [
  1.1950606107711792,
  0,
  0,
  0,
  0,
  1.7925909757614136,
  0,
  0,
  0,
  0,
  -1.0002000331878662,
  -1,
  0,
  0,
  -0.20002000033855438,
  0,
]
`;
